File: crates/solobase-cloudflare/src/tenant.rs
//! Tenant configuration and routing.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use solobase_core::features;

/// Per-tenant configuration stored in KV.
///
/// KV key: `tenant:{subdomain}:config`
///
/// The `config` field holds the tenant's `app.json` — the same format used
/// by the standalone binary.  Feature blocks are present = enabled, absent
/// = disabled.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TenantConfig {
    /// Unique tenant identifier.
    pub id: String,
    /// Subdomain (e.g., `myapp` → `myapp.solobase.app`).
    pub subdomain: String,
    /// Billing plan.
    #[serde(default = "default_plan")]
    pub plan: String,
    /// D1 database UUID for this tenant (created via Cloudflare API).
    /// Each tenant gets its own isolated D1 database.
    #[serde(default)]
    pub db_id: Option<String>,
    /// Worker binding name for this tenant's D1 (e.g., `DB_myapp`).
    /// Falls back to the shared `DB` binding if not set.
    #[serde(default)]
    pub db_binding: Option<String>,
    /// The tenant's app config (same schema as `app.json`).
    #[serde(default)]
    pub config: TenantAppConfig,
    /// Custom WASM block names installed by this tenant.
    /// Each block's compiled .wasm is stored at KV key `tenant:{id}:block:{name}`.
    #[serde(default)]
    pub blocks: Vec<String>,
}

/// Embedded app config — mirrors the standalone `AppConfig` struct but
/// without infrastructure fields that are handled by the Worker (D1, R2, etc.).
///
/// Feature presence rules (same as standalone `app.json`):
/// - Present as `{}` or `true` → enabled
/// - Absent or `false` or `null` → disabled
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TenantAppConfig {
    /// Config format version (default 0 = beta).
    #[serde(default)]
    pub version: u32,

    /// Instance name (informational).
    #[serde(default)]
    pub app: Option<String>,

    // -- Feature blocks (present = enabled, absent = disabled) --

    /// Authentication & user accounts.
    #[serde(default)]
    pub auth: Option<Value>,
    /// Admin panel, IAM, settings, logs, custom tables.
    #[serde(default)]
    pub admin: Option<Value>,
    /// File storage & cloud storage.
    #[serde(default)]
    pub files: Option<Value>,
    /// Product catalog, pricing, purchases, Stripe.
    #[serde(default)]
    pub products: Option<Value>,
    /// Deployment management.
    #[serde(default)]
    pub deployments: Option<Value>,
    /// Legal pages (terms, privacy).
    #[serde(default)]
    pub legalpages: Option<Value>,
    /// User portal (branding, feature toggles).
    #[serde(default)]
    pub userportal: Option<Value>,
}

fn default_plan() -> String {
    "hobby".to_string()
}

// ---------------------------------------------------------------------------
// Feature detection (same logic as standalone AppConfig)
// ---------------------------------------------------------------------------

impl features::FeatureConfig for TenantAppConfig {
    fn auth_enabled(&self) -> bool { features::is_feature_enabled(&self.auth) }
    fn admin_enabled(&self) -> bool { features::is_feature_enabled(&self.admin) }
    fn files_enabled(&self) -> bool { features::is_feature_enabled(&self.files) }
    fn products_enabled(&self) -> bool { features::is_feature_enabled(&self.products) }
    fn deployments_enabled(&self) -> bool { features::is_feature_enabled(&self.deployments) }
    fn legalpages_enabled(&self) -> bool { features::is_feature_enabled(&self.legalpages) }
    fn userportal_enabled(&self) -> bool { features::is_feature_enabled(&self.userportal) }
}

impl TenantAppConfig {
    /// Create a config with all features enabled (the default for new tenants).
    pub fn all_enabled() -> Self {
        let on = Some(Value::Object(Default::default()));
        Self {
            version: 0,
            app: None,
            auth: on.clone(),
            admin: on.clone(),
            files: on.clone(),
            products: on.clone(),
            deployments: on.clone(),
            legalpages: on.clone(),
            userportal: on,
        }
    }
}

File: crates/solobase-cloudflare/worker/tenant.ts
// Tenant resolution and feature gate logic for the TypeScript Cloudflare Worker.
//
// Ports the logic from src/tenant.rs and src/lib.rs (resolve_tenant) to TypeScript.

import type { Env, TenantConfig, TenantAppConfig } from './types';

// ---------------------------------------------------------------------------
// Tenant resolution
// ---------------------------------------------------------------------------

/**
 * Resolve a TenantConfig from the request hostname.
 *
 * - Extracts the subdomain (first label before the first dot).
 * - localhost / 127.0.0.1 returns a hardcoded dev tenant with all features on.
 * - Otherwise looks up `tenant:{subdomain}:config` in the TENANTS KV namespace.
 */
export async function resolveTenant(
  hostname: string,
  env: Env,
): Promise<TenantConfig | null> {
  // Strip port if present (e.g. "localhost:8787")
  const hostNoPort = hostname.split(':')[0];
  const subdomain = hostNoPort.split('.')[0];
  if (!subdomain) return null;

  // `app` subdomain is the platform — never resolve as a tenant.
  // In production, app.solobase.dev is handled by isMarketingHost() in static.ts.
  // In dev mode, localhost is the platform equivalent.
  const isDev = (env.ENVIRONMENT as string) === 'development';
  if (subdomain === 'app') return null;
  if (isDev && (subdomain === 'localhost' || subdomain === '127' || subdomain === 'www')) {
    return {
      id: 'dev',
      subdomain: 'localhost',
      plan: 'hobby',
      db_binding: 'DB',
      config: allEnabledConfig(),
      blocks: [],
    };
  }

  // Look up tenant config in KV
  const kv = env.TENANTS;
  if (!kv) return null;

  const key = `tenant:${subdomain}:config`;
  const raw = await kv.get(key, 'json');
  if (!raw) return null;

  return raw as TenantConfig;
}

// ---------------------------------------------------------------------------
// Feature gates
// ---------------------------------------------------------------------------

/**
 * Feature name to TenantAppConfig field mapping.
 *
 * A feature is considered enabled when its field is present and truthy:
 * - `{}`, `true`, `{...}` -> enabled
 * - `undefined`, `null`, `false` -> disabled
 *
 * System and Profile are always enabled and do not need a feature gate.
 */
export function isFeatureEnabled(
  config: TenantAppConfig,
  feature: string,
): boolean {
  switch (feature) {
    case 'system':
    case 'profile':
      return true;
    case 'auth':
      return isEnabled(config.auth);
    case 'admin':
      return isEnabled(config.admin);
    case 'files':
      return isEnabled(config.files);
    case 'products':
      return isEnabled(config.products);
    case 'deployments':
      return isEnabled(config.deployments);
    case 'legalpages':
      return isEnabled(config.legalpages);
    case 'userportal':
      return isEnabled(config.userportal);
    default:
      return false;
  }
}

/**
 * Same logic as Rust `features::is_feature_enabled`:
 * None / null / false -> disabled; anything else -> enabled.
 */
function isEnabled(val: unknown): boolean {
  if (val === undefined || val === null || val === false) return false;
  return true;
}

// ---------------------------------------------------------------------------
// D1 binding resolution
// ---------------------------------------------------------------------------

/**
 * Get the D1Database binding for a tenant.
 *
 * Uses the tenant's `db_binding` field if set (e.g. `"DB_myapp"`),
 * otherwise falls back to the shared `DB` binding.
 */
export function getD1ForTenant(env: Env, tenant: TenantConfig): D1Database {
  if (tenant.db_binding) {
    const db = env[tenant.db_binding] as D1Database | undefined;
    if (db) return db;
  }
  return env.DB;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Create a TenantAppConfig with all features enabled (dev / new tenant default). */
function allEnabledConfig(): TenantAppConfig {
  return {
    version: 0,
    auth: {},
    admin: {},
    files: {},
    products: {},
    deployments: {},
    legalpages: {},
    userportal: {},
  };
}
